jax/_src/interpreters/mlir.py

```diff
@@ -410,17 +404,49 @@ class ShapePolyLoweringState:
   # from an inner call to a polymorphic Exported.
   uses_dim_vars: bool
 
-  def __init__(self, dim_vars: tuple[str, ...],
-               lowering_platforms: tuple[str, ...]):
-    self.lowering_platforms = lowering_platforms
+  # If the first dimension variable is a platform index argument
+  has_platform_index_argument: bool
+
+  def __init__(self,
+               dim_vars: tuple[str, ...],
+               lowering_platforms: tuple[str, ...] | None):
     self.uses_dim_vars = (len(dim_vars) > 0)
-    if len(lowering_platforms) > 1:
+    if lowering_platforms is not None and len(lowering_platforms) > 1:
       dim_vars = ("platform_index_",) + tuple(dim_vars)
+      self.has_platform_index_argument = True
+    else:
+      self.has_platform_index_argument = False
     self.dim_vars = dim_vars
 
-  @property
-  def has_platform_index_argument(self):
-    return len(self.lowering_platforms) > 1
+
+dataclasses.dataclass(frozen=True)
+class LoweringParameters:
+  # A mapping between primitives and user-defined LoweringRules.
+  # When lowering a primitive, give priorioty to the rule in this map over
+  # existing Jax rules.
+  override_lowering_rules: tuple[tuple[core.Primitive, LoweringRule]] | None = None
```

Comment:
The `|` syntax in type annotations [is 3.10+](https://docs.python.org/3/whatsnew/3.10.html#pep-604-new-type-union-operator), whereas JAX currently [seems to target 3.9](https://github.com/google/jax/blob/173a2701794571f3a102b0dd80bdedb4adbd8a5c/setup.py#L154)+

Reply:
There is already lots of | in JAX, I am not sure how that works with 3.9.

Reply:
Yeah `|` works with 3.9 too.

Reply:
It works in 3.9 if you do `from __future__ import annotations`.